importlab/graph.py

```diff
@@ -189,22 +143,25 @@ def build(self):
 
         assert not self.final, 'Trying to mutate a final graph.'
 
-        # Recursively extract cycles until the graph is cycle-free.
-        while True:
-            try:
-                cycle = Cycle(nx.find_cycle(self.graph))
-                self.extract_cycle(cycle)
-            except nx.NetworkXNoCycle:
+        # Replace each strongly connected component with a single node `NodeSet`
+        for scc in sorted(nx.kosaraju_strongly_connected_components(self.graph),
+                          key=len, reverse=True):
+            if len(scc) == 1:
                 break
 
-        # Now that we have reduced the graph to a tree, we can flatten cycle
-        # nodes into NodeSets
-        def transform_node(node):
-            if isinstance(node, Cycle):
-                return NodeSet(node)
-            else:
-                return node
-        self.graph = nx.relabel_nodes(self.graph, transform_node)
+            ns = NodeSet(scc)
+            self.graph.add_node(ns)
```

Comment:
I'd prefer to keep the code from here to line 163 in a separate `extract_cycle` method - in general, smaller methods are easier to test/maintain, and in this particular case, I don't see any advantage to inlining the method.

Reply:
Since a strongly connected component is not a cycle anymore, I extracted this part of code in a function called `shrink_to_node`.